File: westcoast/src/main/java/frc/robot/definition/Definition.java
package frc.robot.definition;

import frc.subsystems.DriveTrainSubsystem;
import frc.subsystems.IntakeSubsystem;
import frc.subsystems.MagazineSubsystem;
import frc.subsystems.ShooterSubsystem;
import org.jetbrains.annotations.NotNull;

public class Definition {
    @NotNull
    public final String name;
    @NotNull
    public final Input input;
    @NotNull
    public final Motors motors;
    @NotNull
    public final Pneumatics pneumatics;
    @NotNull
    public final Sensors sensors;
    @NotNull
    public final Subsystems subsystems;
    @NotNull
    public final Parameters parameters;
    @NotNull
    public final Controllers controllers;

    public Definition(
            @NotNull String name,
            @NotNull Input input,
            @NotNull Motors motors,
            @NotNull Pneumatics pneumatics,
            @NotNull Sensors sensors,
            @NotNull Subsystems subsystems,
            @NotNull Parameters parameters,
            @NotNull Controllers controllers
    ) {
        this.name = name;
        this.input = input;
        this.motors = motors;
        this.pneumatics = pneumatics;
        this.sensors = sensors;
        this.subsystems = subsystems;
        this.parameters = parameters;
        this.controllers = controllers;
    }

    public static Definition definition(
            @NotNull String name,
            @NotNull Input input,
            @NotNull Motors motors,
            @NotNull Pneumatics pneumatics,
            @NotNull Sensors sensors,
            @NotNull Parameters parameters,
            @NotNull Controllers controllers
    ) {
        return new Definition(
                name,
                input,
                motors,
                pneumatics,
                sensors,
                subsystems(motors, pneumatics, sensors),
                parameters,
                controllers
        );
    }

    @NotNull
    private static Subsystems subsystems(
            @NotNull Motors motors,
            @NotNull Pneumatics pneumatics,
            @NotNull Sensors sensors
    ) {
        return new Subsystems(
                new DriveTrainSubsystem(
                        motors.driveLeft.lead,
                        motors.driveRight.lead,
                        sensors.drive.leftEncoder,
                        sensors.drive.rightEncoder,
                        sensors.drive.gyro,
                        sensors.drive.vision
                ),
                new IntakeSubsystem(
                        pneumatics.intakeDeploy,
                        motors.intake.main
                ),
                new ShooterSubsystem(
                        motors.shooter.lead
                ),
                new MagazineSubsystem(
                        motors.magazine.index,
                        motors.magazine.feeder,
                        motors.magazine.reject,
                        sensors.magazine.index,
                        sensors.magazine.feeder,
                        sensors.magazine.reject,
                        sensors.magazine.colorSensor
                )
        );
    }

}

File: westcoast/src/main/kotlin/frc/robot/definition/Definition.kt
package frc.robot.definition

import frc.subsystems.*

data class Definition(
    val name: String,
    val input: Input,
    val motors: Motors,
    val pneumatics: Pneumatics,
    val sensors: Sensors,
    val parameters: Parameters,
    val controllers: Controllers,
    val subsystems: Subsystems = Subsystems(
        driveTrain = DriveTrainSubsystem(
            leftMotor = motors.driveLeft.lead,
            rightMotor = motors.driveRight.lead,
            leftEncoder = sensors.drive.leftEncoder,
            rightEncoder = sensors.drive.rightEncoder,
            gyro = sensors.drive.gyro,
            vision = sensors.drive.vision
        ),
        intake = IntakeSubsystem(
            deploy = pneumatics.intakeDeploy,
            intakeMotor = motors.intake.main
        ),
        shooter = ShooterSubsystem(
            shooterMotor = motors.shooter.lead,
            configurator = { controllers.shooterConfig(this) }
        ),
        magazine = MagazineSubsystem(
            indexMotor = motors.magazine.index,
            feederMotor = motors.magazine.feeder,
            rejectMotor = motors.magazine.reject,
            indexSensor = sensors.magazine.index,
            feederSensor = sensors.magazine.feeder,
            rejectSensor = sensors.magazine.reject,
            colorSensor = sensors.magazine.colorSensor
        ),
        climber = ClimberSubsystem(
            climberMotor = motors.climber.lead,
            climberConfigDeploy = { controllers.climber.deploy(this) },
            climberConfigClimb = { controllers.climber.climb(this) }
        )
    )
)